src/llama_index_alloydb_pg/async_chat_store.py

```diff
@@ -0,0 +1,295 @@
+        """
+        insert_query = f"""
+                INSERT INTO "{self._schema_name}"."{self._table_name}" (key, message)
+                VALUES (:key, :message);"""
+        params = {"key": key, "message": json.dumps(message.dict())}
+
+        await self.__aexecute_query(insert_query, params)
+
+    async def adelete_messages(self, key: str) -> Optional[List[ChatMessage]]:
+        """Asynchronously deletes the chat messages associated with a specific key.
+
+        Args:
+            key (str): A unique identifier for the chat whose messages are to be deleted.
+
+        Returns:
+            Optional[List[ChatMessage]]: A list of `ChatMessage` objects that were deleted, or `None` if no messages
+            were associated with the key or could be deleted.
+        """
+        query = f"""DELETE FROM "{self._schema_name}"."{self._table_name}" WHERE key = '{key}' RETURNING *; """
+        results = await self.__afetch_query(query)
+        if results:
+            return [
+                ChatMessage.model_validate(result.get("message")) for result in results
+            ]
+        return None
+
+    async def adelete_message(self, key: str, idx: int) -> Optional[ChatMessage]:
+        """Asynchronously deletes a specific chat message by index from the messages associated with a given key.
+
+        Args:
+            key (str): A unique identifier for the chat whose messages are to be deleted.
+            idx (int): The index of the `ChatMessage` to be deleted from the list of messages.
+
+        Returns:
+            Optional[ChatMessage]: The `ChatMessage` object that was deleted, or `None` if no message
+            was associated with the key or could be deleted.
+        """
+        query = f"""SELECT * from "{self._schema_name}"."{self._table_name}" WHERE key = '{key}' ORDER BY id;"""
+        results = await self.__afetch_query(query)
+        if results:
+            if idx >= len(results):
+                return None
+            id_to_be_deleted = results[idx].get("id")
+            delete_query = f"""DELETE FROM "{self._schema_name}"."{self._table_name}" WHERE id = '{id_to_be_deleted}' RETURNING *;"""
+            result = await self.__afetch_query(delete_query)
+            result = result[0]
+            if result:
+                return ChatMessage.model_validate(result.get("message"))
+            return None
+        return None
+
+    async def adelete_last_message(self, key: str) -> Optional[ChatMessage]:
+        """Asynchronously deletes the last chat message associated with a given key.
+
+        Args:
+            key (str): A unique identifier for the chat whose message is to be deleted.
+
+        Returns:
+            Optional[ChatMessage]: The `ChatMessage` object that was deleted, or `None` if no message
+            was associated with the key or could be deleted.
+        """
+        query = f"""SELECT * from "{self._schema_name}"."{self._table_name}" WHERE key = '{key}' ORDER BY id;"""
```

Comment:
nit: We could pull the last id like `f"""SELECT * from "{self._schema_name}"."{self._table_name}" WHERE key = '{key}' ORDER BY id DESC LIMIT 1;"""`. We might want to ask a postgres expert if there is an improvement in latency using Postgres to select the last element or using python to select the last element from the list.